File: src/haruhiloop_cli/systems/memory.py
from __future__ import annotations

from haruhiloop_cli.models import GameState, clamp


def apply_memory_residue(state: GameState) -> None:
    clue_gain = 0
    sync_gain = 0
    if "truth_shared" in state.flags:
        clue_gain += 1
        sync_gain += 1
    if "homework_done" in state.flags:
        sync_gain += 1
    if "hope_signal" in state.flags:
        clue_gain += 1
    if state.closed_space_stage > 0:
        sync_gain += 1

    residue = state.memory_residue
    residue["clue_efficiency"] = min(5, residue.get("clue_efficiency", 0) + clue_gain)
    residue["sync_recovery"] = min(5, residue.get("sync_recovery", 0) + sync_gain)


def apply_residue_decay(state: GameState) -> None:
    residue = state.memory_residue
    residue["clue_efficiency"] = max(0, residue.get("clue_efficiency", 0) - 1)
    residue["sync_recovery"] = max(0, residue.get("sync_recovery", 0) - 1)


def inject_residue_bonus(state: GameState) -> None:
    residue = state.memory_residue
    if residue.get("clue_efficiency", 0) > 0:
        state.clue_points = clamp(state.clue_points + 1)
    if residue.get("sync_recovery", 0) > 0:
        state.crew_sync = clamp(state.crew_sync + 1)

File: web/src/domain/systems/memory.ts
// 与 src/haruhiloop_cli/systems/memory.py 1:1 对齐。

import { GameState } from "../models";
import { clamp } from "../../infrastructure/clamp";

export function applyMemoryResidue(state: GameState): void {
  let clueGain = 0;
  let syncGain = 0;
  if (state.flags.has("truth_shared")) {
    clueGain += 1;
    syncGain += 1;
  }
  if (state.flags.has("homework_done")) syncGain += 1;
  if (state.flags.has("hope_signal")) clueGain += 1;
  if (state.closed_space_stage > 0) syncGain += 1;

  const r = state.memory_residue;
  r["clue_efficiency"] = Math.min(5, (r["clue_efficiency"] ?? 0) + clueGain);
  r["sync_recovery"] = Math.min(5, (r["sync_recovery"] ?? 0) + syncGain);
}

export function applyResidueDecay(state: GameState): void {
  const r = state.memory_residue;
  r["clue_efficiency"] = Math.max(0, (r["clue_efficiency"] ?? 0) - 1);
  r["sync_recovery"] = Math.max(0, (r["sync_recovery"] ?? 0) - 1);
}

export function injectResidueBonus(state: GameState): void {
  const r = state.memory_residue;
  if ((r["clue_efficiency"] ?? 0) > 0) {
    state.clue_points = clamp(state.clue_points + 1);
  }
  if ((r["sync_recovery"] ?? 0) > 0) {
    state.crew_sync = clamp(state.crew_sync + 1);
  }
}
